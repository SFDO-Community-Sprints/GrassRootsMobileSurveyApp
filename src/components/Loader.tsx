import React from 'react';
import { StyleSheet, View, Modal, ActivityIndicator, Text } from 'react-native';
import { APP_THEME, APP_FONTS } from '../constants';

type LoaderProps = {
  loading: boolean;
  message?: string;
};

function Loader(props: LoaderProps) {
  const { loading, message } = props;
  return (
    <Modal transparent visible={loading}>
      <View style={styles.modalBackground}>
        <ActivityIndicator animating={loading} size="large" hidesWhenStopped color={APP_THEME.APP_BASE_COLOR} />
        <Text style={styles.loadingTitleStyle}> {message} </Text>
      </View>
    </Modal>
  );
}

const styles = StyleSheet.create({
  modalBackground: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: APP_THEME.APP_WHITE,
    alignItems: 'center',
    opacity: 0.8,
  },
  loadingTitleStyle: {
    fontFamily: APP_FONTS.FONT_REGULAR,
    fontSize: 36,
    color: APP_THEME.APP_BASE_COLOR,
    paddingTop: 10,
  },
});

export { Loader };
