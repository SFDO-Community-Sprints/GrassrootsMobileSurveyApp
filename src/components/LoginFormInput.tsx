import React from 'react';
import { StyleSheet } from 'react-native';
import { Input } from '@rneui/themed';
import { APP_THEME, APP_FONTS } from '../constants';

type LoginFormInputProps = {
  onChangeText(value: string): void;
  value: string;
  label: string;
  placeholder: string;
  icon: string;
  isSecure?: boolean;
  errorMessage: string;
  keyboardType?: 'email-address' | 'default';
};

function LoginFormInput(props: LoginFormInputProps) {
  return (
    <Input
      onChangeText={value => {
        props.onChangeText(value);
      }}
      value={props.value}
      label={props.label}
      placeholder={props.placeholder}
      leftIcon={{ type: 'material-community', name: props.icon, color: APP_THEME.APP_LIGHT_FONT_COLOR }}
      errorStyle={styles.errorStyle}
      labelStyle={styles.font}
      inputStyle={styles.font}
      keyboardType={props.keyboardType}
      errorMessage={props.errorMessage}
      secureTextEntry={props.isSecure}
    />
  );
}

const styles = StyleSheet.create({
  font: {
    fontFamily: APP_FONTS.FONT_REGULAR,
  },
  errorStyle: {
    color: APP_THEME.APP_ERROR_COLOR,
    fontFamily: APP_FONTS.FONT_REGULAR,
  },
});

export { LoginFormInput };
