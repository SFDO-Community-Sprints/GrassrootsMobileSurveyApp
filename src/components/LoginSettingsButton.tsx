import React from 'react';
import { Icon } from '@rneui/themed';
import { StackNavigationProp } from '@react-navigation/stack';

import { APP_THEME } from '../constants';

import { StackParamList } from '../Router';
type LoginScreenNavigationProp = StackNavigationProp<StackParamList, 'SurveyEditor'>;

type LoginSettingsButtonProps = {
  navigation: LoginScreenNavigationProp;
};

export function LoginSettingsButton(props: LoginSettingsButtonProps) {
  return (
    <Icon
      accessibilityRole="button"
      accessible
      iconStyle={{ padding: 10 }}
      name="settings"
      size={22}
      color={APP_THEME.APP_BASE_COLOR}
      type="material"
      onPress={() => {
        props.navigation.navigate('LoginSettings');
      }}
    />
  );
}
