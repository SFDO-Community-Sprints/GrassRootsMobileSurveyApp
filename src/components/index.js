export * from './TextHeading';
export * from './TextSubheading';
export * from './ListItem';
export * from './Loader';
export * from './TextIcon';
export * from './LoginFormInput';
