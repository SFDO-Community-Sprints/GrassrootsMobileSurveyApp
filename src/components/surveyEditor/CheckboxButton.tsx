import React from 'react';
import { View, TouchableOpacity, StyleSheet } from 'react-native';
import { Icon } from '@rneui/themed';
import { APP_THEME } from '../../constants';

type CheckboxButtonPropType = {
  title: JSX.Element;
  onPress(): void;
  selected: boolean;
  disabled?: boolean;
};

/**
 * Checkbox form input component for survey
 */
function CheckboxButton(props: CheckboxButtonPropType) {
  const { title, onPress, selected, disabled } = props;

  return (
    <TouchableOpacity style={styles.containerStyle} disabled={disabled} onPress={onPress}>
      <Icon
        name={selected ? 'check-box' : 'check-box-outline-blank'}
        size={30}
        color={APP_THEME.APP_BASE_COLOR}
        type="material"
      />
      <View style={{ marginBottom: -5 }}>{title}</View>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  containerStyle: {
    padding: 10,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

export { CheckboxButton };
