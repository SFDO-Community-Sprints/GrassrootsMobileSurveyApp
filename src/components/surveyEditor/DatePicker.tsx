import React, { useState } from 'react';
import { View, StyleSheet, Text, TouchableOpacity } from 'react-native';
import { Icon } from '@rneui/themed';
import DateTimePickerModal from 'react-native-modal-datetime-picker';

import { APP_THEME, APP_FONTS } from '../../constants';
import { useLocalizationContext } from '../../context/localizationContext';
import { formatISOStringToCalendarDateString } from '../../utility/date';

type DatePickerPropType = {
  title: string;
  mode: 'date' | 'time' | 'datetime';
  fieldLabel: JSX.Element;
  value: string;
  onValueChange(date: string): void;
  disabled?: boolean;
};

function DatePicker(props: DatePickerPropType) {
  const [isDatePickerVisible, setIsDatePickerVisible] = useState(false);

  const { t } = useLocalizationContext();

  const { onValueChange, value, disabled, fieldLabel, mode } = props;
  const { container, innerContainer, buttonStyle, valueLabel, iconView, placeholderLabel } = styles;

  const toSalesforceTimeValue = (date: Date): string => {
    return `${date.toTimeString().split(' ')[0]}.000Z`;
  };

  const toDateValue = (value: string): Date => {
    if (value && mode === 'date') {
      return new Date(value);
    } else if (value && mode === 'time') {
      return new Date(`1990-01-01T${value}`);
    }
    return undefined;
  };

  const displayValue = (value: string): string => {
    if (value && mode === 'date') {
      return formatISOStringToCalendarDateString(value);
    } else if (value && mode === 'time') {
      return value.substring(0, 5);
    }
    return t('SELECT');
  };

  return (
    <View style={container}>
      {fieldLabel}
      <View style={innerContainer}>
        <TouchableOpacity
          style={buttonStyle}
          disabled={disabled}
          onPress={() => {
            setIsDatePickerVisible(true);
          }}
        >
          <Text style={value ? valueLabel : placeholderLabel}>{displayValue(value)}</Text>
          <View style={iconView}>
            <Icon
              name={mode === 'time' ? 'clockcircleo' : 'calendar'}
              size={18}
              color={APP_THEME.APP_BASE_FONT_COLOR}
              type="antdesign"
            />
          </View>
        </TouchableOpacity>
      </View>
      <DateTimePickerModal
        confirmTextIOS={t('CONFIRM')}
        cancelTextIOS={t('CANCEL')}
        // headerTextIOS={title}
        date={toDateValue(value)}
        mode={mode}
        isVisible={isDatePickerVisible}
        onConfirm={date => {
          if (mode !== 'time') {
            onValueChange(date.toISOString());
          } else {
            onValueChange(toSalesforceTimeValue(date));
          }
          setIsDatePickerVisible(false);
        }}
        onCancel={() => setIsDatePickerVisible(false)}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: { padding: 10 },
  innerContainer: {
    minHeight: 40,
    borderWidth: 1,
    borderRadius: 4,
    borderColor: APP_THEME.APP_BORDER_COLOR,
    alignItems: 'center',
    paddingLeft: 10,
    width: '100%',
  },
  buttonStyle: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  valueLabel: {
    flex: 1,
    fontSize: 16,
    fontFamily: APP_FONTS.FONT_REGULAR,
    color: APP_THEME.APP_DARK_FONT_COLOR,
  },
  iconView: {
    right: 0,
    width: 40,
    height: 40,
    justifyContent: 'center',
  },
  placeholderLabel: {
    flex: 1,
    fontSize: 16,
    color: APP_THEME.APP_LIGHT_FONT_COLOR,
    fontFamily: APP_FONTS.FONT_REGULAR,
  },
});

export { DatePicker };
