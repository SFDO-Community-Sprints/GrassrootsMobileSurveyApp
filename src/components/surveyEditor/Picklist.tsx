import React, { useState, useEffect } from 'react';
import { View, StyleSheet } from 'react-native';
import RNPickerSelect from 'react-native-picker-select';

import { APP_THEME } from '../../constants';
import { getPicklistValues } from '../../services/database/metadata';

type PicklistPropType = {
  fieldName: string;
  fieldLabel: JSX.Element;
  value: string;
  onValueChange(value: string): void;
  disabled?: boolean;
  hideNone?: boolean;
};

function Picklist(props: PicklistPropType) {
  const [options, setOptions] = useState([]);

  useEffect(() => {
    const setPicklistValues = async () => {
      const storedOptions = await getPicklistValues(props.fieldName);
      setOptions(storedOptions.map(o => ({ label: o.label, value: o.value })));
    };
    setPicklistValues();
  }, []);

  const { onValueChange, value, disabled, fieldLabel } = props;

  return (
    <View
      style={{
        padding: 10,
        width: '100%',
      }}
    >
      {fieldLabel}
      {options.length > 0 && (
        <RNPickerSelect
          disabled={disabled}
          value={value}
          items={options}
          style={pickerSelectStyles}
          // dropDownStyle={{ backgroundColor: APP_THEME.APP_WHITE }}
          // placeholder={{}}
          onValueChange={value => {
            onValueChange(value);
          }}
        />
      )}
    </View>
  );
}

const pickerSelectStyles = StyleSheet.create({
  inputIOS: {
    paddingVertical: 12,
    paddingHorizontal: 10,
    borderRadius: 4,
    borderWidth: 1,
    borderColor: APP_THEME.APP_BORDER_COLOR,
    color: 'black',
  },
  inputAndroid: {
    paddingHorizontal: 10,
    paddingVertical: 8,
    borderRadius: 4,
    borderWidth: 1,
    borderColor: APP_THEME.APP_BORDER_COLOR,
    color: 'black',
    paddingRight: 30, // to ensure the text is never behind the icon
  },
});

export { Picklist };
