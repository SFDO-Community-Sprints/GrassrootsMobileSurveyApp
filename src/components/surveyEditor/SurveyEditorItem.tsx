import React, { memo, useContext } from 'react';
import { Text, View, StyleSheet } from 'react-native';
import { TextInput, CheckboxButton, DatePicker, Picklist, Lookup } from '.';
import { StackNavigationProp } from '@react-navigation/stack';

import { useSelector, useDispatch } from '../../state/surveyEditorState';
import LocalizationContext from '../../context/localizationContext';
import { StackParamList } from '../../Router';
import { APP_THEME, APP_FONTS, L10N_PREFIX, SYNC_STATUS_SYNCED } from '../../constants';

type SurveyEditorNavigationProp = StackNavigationProp<StackParamList, 'SurveyEditor'>;

type SurveyItemProps = {
  navigation: SurveyEditorNavigationProp;
  title: string;
  name: string;
  type: string;
  required: boolean;
};

function SurveyEditorItem({ navigation, title, name, type, required }: SurveyItemProps) {
  const value = useSelector(state => state.survey[name]);
  const _syncStatus = useSelector(state => state.survey._syncStatus);
  const disabled = _syncStatus === SYNC_STATUS_SYNCED;
  const dispatchSurvey = useDispatch();
  const { t } = useContext(LocalizationContext);

  const onValueChange = value => {
    dispatchSurvey({ type: 'UPDATE', field: { name: name, value } });
  };

  const fieldLabel = () => {
    return (
      <View style={styles.labelContainer}>
        {required && <Text style={{ color: APP_THEME.APP_ERROR_COLOR }}>*</Text>}
        <Text style={styles.labelStyle}>{t(`${L10N_PREFIX.PageLayoutItem}${name}`)}</Text>
      </View>
    );
  };

  const renderItem = () => {
    switch (type) {
      case 'string':
        return <TextInput title={fieldLabel()} onValueChange={onValueChange} value={value} disabled={disabled} />;
      case 'email':
        return (
          <TextInput
            title={fieldLabel()}
            onValueChange={onValueChange}
            value={value}
            disabled={disabled}
            keyboardType="email-address"
          />
        );
      case 'textarea':
        return (
          <TextInput title={fieldLabel()} onValueChange={onValueChange} value={value} multiline disabled={disabled} />
        );
      case 'double':
        return (
          <TextInput
            title={fieldLabel()}
            onValueChange={onValueChange}
            value={value?.toString()}
            keyboardType="numeric"
            disabled={disabled}
          />
        );
      case 'currency':
        return (
          <TextInput
            title={fieldLabel()}
            onValueChange={onValueChange}
            value={value?.toString()}
            keyboardType="numeric"
            disabled={disabled}
          />
        );
      case 'boolean':
        return (
          <CheckboxButton
            title={fieldLabel()}
            onPress={() => dispatchSurvey({ type: 'UPDATE', field: { name: name, value: !value } })}
            selected={value}
            disabled={disabled}
          />
        );
      case 'date':
        return (
          <DatePicker
            mode="date"
            title={title}
            fieldLabel={fieldLabel()}
            onValueChange={onValueChange}
            value={value}
            disabled={disabled}
          />
        );
      case 'time':
        return (
          <DatePicker
            mode="time"
            title={title}
            fieldLabel={fieldLabel()}
            onValueChange={onValueChange}
            value={value}
            disabled={disabled}
          />
        );
      case 'picklist':
        return (
          <Picklist
            onValueChange={onValueChange}
            value={value}
            fieldName={name}
            fieldLabel={fieldLabel()}
            disabled={disabled}
          />
        );
      case 'phone':
        return <TextInput title={fieldLabel()} onValueChange={onValueChange} value={value} keyboardType="phone-pad" />;
      case 'reference':
        return (
          <Lookup
            title={title}
            fieldName={name}
            fieldLabel={fieldLabel()}
            navigation={navigation}
            value={value}
            disabled={disabled}
          />
        );
      default:
        return (
          <Text>
            {title} ({type})
          </Text>
        );
    }
  };

  return (
    <View
      style={{
        // flex: 1,
        alignItems: 'flex-start',
        backgroundColor: APP_THEME.APP_WHITE,
      }}
    >
      {renderItem()}
    </View>
  );
}

const styles = StyleSheet.create({
  labelContainer: {
    flexDirection: 'row',
    paddingBottom: 5,
  },
  labelStyle: {
    fontSize: 14,
    paddingLeft: 5,
    color: APP_THEME.APP_LIGHT_FONT_COLOR,
    fontFamily: APP_FONTS.FONT_BOLD,
  },
});

export default memo(SurveyEditorItem);
