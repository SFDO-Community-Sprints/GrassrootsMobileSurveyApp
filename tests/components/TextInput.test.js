import React from 'react';
import { render } from '@testing-library/react-native';

import { TextInput } from '../../src/components/surveyEditor';

describe('<TextInput />', () => {
  it('Rendering', () => {
    const screen = render(<TextInput title="Test" />);
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
