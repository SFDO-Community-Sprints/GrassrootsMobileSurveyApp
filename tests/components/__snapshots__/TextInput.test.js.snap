// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> Rendering 1`] = `
<View
  style={
    Object {
      "padding": 10,
      "paddingHorizontal": 10,
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#86939e",
        "fontSize": 16,
        "fontWeight": "bold",
      }
    }
  >
    Test
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "#dedcdb",
        "borderRadius": 4,
        "borderWidth": 1,
        "flexDirection": "row",
        "paddingLeft": 10,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      editable={true}
      keyboardType="default"
      multiline={false}
      numberOfLines={1}
      placeholderTextColor="#86939e"
      rejectResponderTermination={true}
      style={
        Object {
          "alignSelf": "center",
          "color": "#242424",
          "flex": 1,
          "fontFamily": "SalesforceSans-Regular",
          "fontSize": 16,
          "minHeight": 40,
        }
      }
      testID="RNE__Input__text-input"
      underlineColorAndroid="transparent"
    />
  </View>
  <Text
    style={
      Object {
        "color": "#ff190c",
        "fontSize": 12,
        "height": 0,
        "margin": 0,
        "padding": 0,
      }
    }
  />
</View>
`;
